File: backend/presentation/reports.py
from abc import ABC, abstractmethod
from backend.dal.dbconfig import dbconfig
import json
from datetime import datetime

class GenerateReport(ABC):
    @abstractmethod
    def getData(self):
        # Execute query and fetch data
        pass

    @abstractmethod
    def processData(self):
        # Logic to process self.data
        pass

    @abstractmethod
    def outputData(self):
        # Logic to output the processed data
        pass

class EnrollmentStatisticsReport(GenerateReport):
    def __init__(self, dept_id=None, semester=None):
        self.db = dbconfig()
        self.conn = self.db.get_db_connection()
        self.cursor = self.conn.cursor()
        self.dept_id = dept_id
        self.semester = semester

    def getData(self):
        # Execute query and fetch enrollment statistics data
        if self.dept_id:
            query = """
            SELECT C.course_id, C.courseName, dept.deptName, C.availableSeats, C.capacity,
                   (C.capacity - C.availableSeats) as filled_seats,
                   CONCAT(U.firstName, ' ', U.lastName) as instructor_name
            FROM Course AS C
            JOIN Department as dept ON C.dept_Id = dept.dept_Id
            JOIN Users as U ON C.facultyMem_Id = U.user_id
            WHERE dept.dept_Id = %s
            ORDER BY C.courseName;
            """
            self.cursor.execute(query, (self.dept_id,))
        else:
            query = """
            SELECT C.course_id, C.courseName, dept.deptName, C.availableSeats, C.capacity,
                   (C.capacity - C.availableSeats) as filled_seats,
                   CONCAT(U.firstName, ' ', U.lastName) as instructor_name
            FROM Course AS C
            JOIN Department as dept ON C.dept_Id = dept.dept_Id
            JOIN Users as U ON C.facultyMem_Id = U.user_id
            ORDER BY dept.deptName, C.courseName;
            """
            self.cursor.execute(query)
        courses = self.cursor.fetchall()
        return courses

    def processData(self):
        # Logic to process enrollment statistics data
        courses = self.getData()
        processed_data = []
        for course in courses:
            course_id, course_name, dept_name, available_seats, capacity, filled_seats, instructor = course
            utilization_percentage = (filled_seats / capacity * 100) if capacity > 0 else 0
            
            processed_data.append({
                "course_id": course_id,
                "courseName": course_name,
                "department": dept_name,
                "instructor": instructor,
                "availableSeats": available_seats,
                "capacity": capacity,
                "filledSeats": filled_seats,
                "utilizationPercentage": round(utilization_percentage, 2),
                "status": "Full" if available_seats == 0 else "Open"
            })
        return processed_data

    def outputData(self):
        # Logic to output the processed enrollment statistics data
        departmentCourses = self.processData()
        return departmentCourses
class FacultyWorkloadReport(GenerateReport):
    def __init__(self, faculty_id=None):
        self.db = dbconfig()
        self.conn = self.db.get_db_connection()
        self.cursor = self.conn.cursor()
        self.faculty_id = faculty_id

    def getData(self):
        # Execute query and fetch faculty workload data
        if self.faculty_id:
            query = """
            SELECT f.facultyMem_Id, CONCAT(u.firstName, ' ', u.lastName) as facultyName, 
                   d.deptName, COUNT(C.course_Id) AS numberOfCourses,
                   SUM(C.capacity - C.availableSeats) AS totalEnrolledStudents,
                   AVG(C.capacity - C.availableSeats) AS avgStudentsPerCourse
            FROM FacultyStaff AS f
            JOIN Users as u ON f.facultyMem_Id = u.user_id
            LEFT JOIN Course AS C ON f.facultyMem_Id = C.facultyMem_Id
            LEFT JOIN Department AS d ON C.dept_Id = d.dept_Id
            WHERE f.facultyMem_Id = %s
            GROUP BY f.facultyMem_Id, u.firstName, u.lastName, d.deptName;
            """
            self.cursor.execute(query, (self.faculty_id,))
        else:
            query = """
            SELECT f.facultyMem_Id, CONCAT(u.firstName, ' ', u.lastName) as facultyName, 
                   d.deptName, COUNT(C.course_Id) AS numberOfCourses,
                   SUM(C.capacity - C.availableSeats) AS totalEnrolledStudents,
                   AVG(C.capacity - C.availableSeats) AS avgStudentsPerCourse
            FROM FacultyStaff AS f
            JOIN Users as u ON f.facultyMem_Id = u.user_id
            LEFT JOIN Course AS C ON f.facultyMem_Id = C.facultyMem_Id
            LEFT JOIN Department AS d ON C.dept_Id = d.dept_Id
            GROUP BY f.facultyMem_Id, u.firstName, u.lastName, d.deptName
            ORDER BY d.deptName, u.lastName;
            """
            self.cursor.execute(query)
        results = self.cursor.fetchall()
        return results

    def processData(self):
        # Process the data to a list of dicts
        results = self.getData()
        processed = []
        for row in results:
            processed.append({
                "facultyId": row[0],
                "facultyName": row[1],
                "department": row[2] or "Unassigned",
                "numberOfCourses": row[3] or 0,
                "totalEnrolledStudents": row[4] or 0,
                "avgStudentsPerCourse": round(row[5] or 0, 2)
            })
        return processed

    def outputData(self):
        # Output the processed faculty workload data
        return self.processData()


class CoursePopularityReport(GenerateReport):
    def __init__(self, semester=None, limit=10):
        self.db = dbconfig()
        self.conn = self.db.get_db_connection()
        self.cursor = self.conn.cursor()
        self.semester = semester
        self.limit = limit

    def getData(self):
        # Get course popularity data based on enrollment
        query = """
        SELECT C.course_id, C.courseName, dept.deptName,
               CONCAT(U.firstName, ' ', U.lastName) as instructor_name,
               C.capacity, C.availableSeats,
               (C.capacity - C.availableSeats) as enrolled_count,
               ROUND(((C.capacity - C.availableSeats) / C.capacity) * 100, 2) as popularity_percentage
        FROM Course AS C
        JOIN Department as dept ON C.dept_Id = dept.dept_Id
        JOIN Users as U ON C.facultyMem_Id = U.user_id
        WHERE C.capacity > 0
        ORDER BY enrolled_count DESC, popularity_percentage DESC
        LIMIT %s;
        """
        self.cursor.execute(query, (self.limit,))
        results = self.cursor.fetchall()
        return results

    def processData(self):
        results = self.getData()
        processed = []
        for i, row in enumerate(results, 1):
            processed.append({
                "rank": i,
                "course_id": row[0],
                "courseName": row[1],
                "department": row[2],
                "instructor": row[3],
                "capacity": row[4],
                "availableSeats": row[5],
                "enrolledCount": row[6],
                "popularityPercentage": row[7]
            })
        return processed

    def outputData(self):
        return self.processData()


class HighCapacityCoursesReport(GenerateReport):
    def __init__(self, department_name=None, threshold_percentage=90):
        self.db = dbconfig()
        self.conn = self.db.get_db_connection()
        self.cursor = self.conn.cursor()
        self.department_name = department_name
        self.threshold_percentage = threshold_percentage

    def getData(self):
        # Get courses above the capacity threshold
        if self.department_name:
            query = """
            SELECT C.course_id, C.courseName, dept.deptName,
                   CONCAT(U.firstName, ' ', U.lastName) as instructor_name,
                   C.capacity, C.availableSeats,
                   (C.capacity - C.availableSeats) as enrolled_count,
                   ROUND(((C.capacity - C.availableSeats) / C.capacity) * 100, 2) as utilization_percentage
            FROM Course AS C
            JOIN Department as dept ON C.dept_Id = dept.dept_Id
            JOIN Users as U ON C.facultyMem_Id = U.user_id
            WHERE dept.deptName LIKE %s
            AND C.capacity > 0
            AND ((C.capacity - C.availableSeats) / C.capacity) * 100 >= %s
            ORDER BY utilization_percentage DESC;
            """
            self.cursor.execute(query, (f"%{self.department_name}%", self.threshold_percentage))
        else:
            query = """
            SELECT C.course_id, C.courseName, dept.deptName,
                   CONCAT(U.firstName, ' ', U.lastName) as instructor_name,
                   C.capacity, C.availableSeats,
                   (C.capacity - C.availableSeats) as enrolled_count,
                   ROUND(((C.capacity - C.availableSeats) / C.capacity) * 100, 2) as utilization_percentage
            FROM Course AS C
            JOIN Department as dept ON C.dept_Id = dept.dept_Id
            JOIN Users as U ON C.facultyMem_Id = U.user_id
            WHERE C.capacity > 0
            AND ((C.capacity - C.availableSeats) / C.capacity) * 100 >= %s
            ORDER BY utilization_percentage DESC;
            """
            self.cursor.execute(query, (self.threshold_percentage,))
        results = self.cursor.fetchall()
        return results

    def processData(self):
        results = self.getData()
        processed = []
        for row in results:
            processed.append({
                "course_id": row[0],
                "courseName": row[1],
                "department": row[2],
                "instructor": row[3],
                "capacity": row[4],
                "availableSeats": row[5],
                "enrolledCount": row[6],
                "utilizationPercentage": row[7],
                "status": "Critical" if row[7] >= 95 else "High"
            })
        return processed

    def outputData(self):
        return self.processData()


class DepartmentAnalyticsReport(GenerateReport):
    def __init__(self, semester=None):
        self.db = dbconfig()
        self.conn = self.db.get_db_connection()
        self.cursor = self.conn.cursor()
        self.semester = semester

    def getData(self):
        # Get comprehensive department analytics
        query = """
        SELECT 
            dept.dept_Id,
            dept.deptName,
            COUNT(C.course_id) as total_courses,
            SUM(C.capacity) as total_capacity,
            SUM(C.capacity - C.availableSeats) as total_enrolled,
            SUM(C.availableSeats) as total_available,
            ROUND(AVG((C.capacity - C.availableSeats) / C.capacity * 100), 2) as avg_utilization,
            COUNT(DISTINCT C.facultyMem_Id) as faculty_count
        FROM Department dept
        LEFT JOIN Course C ON dept.dept_Id = C.dept_Id
        WHERE C.capacity > 0
        GROUP BY dept.dept_Id, dept.deptName
        ORDER BY avg_utilization DESC;
        """
        self.cursor.execute(query)
        results = self.cursor.fetchall()
        return results

    def processData(self):
        results = self.getData()
        processed = []
        for row in results:
            processed.append({
                "departmentId": row[0],
                "departmentName": row[1],
                "totalCourses": row[2] or 0,
                "totalCapacity": row[3] or 0,
                "totalEnrolled": row[4] or 0,
                "totalAvailable": row[5] or 0,
                "avgUtilization": row[6] or 0,
                "facultyCount": row[7] or 0
            })
        return processed

    def outputData(self):
        return self.processData()


class ReportExporter:
    @staticmethod
    def export_to_json(data, filename=None):
        """Export report data to JSON format"""
        if not filename:
            filename = f"report_{datetime.now().strftime('%Y%m%d_%H%M%S')}.json"
        
        export_data = {
            "generated_at": datetime.now().isoformat(),
            "data": data
        }
        
        return json.dumps(export_data, indent=2, default=str)
    
    @staticmethod
    def export_to_html_table(data, title="Report"):
        """Export report data to HTML table format"""
        if not data:
            return f"<h2>{title}</h2><p>No data available</p>"
        
        # Get column headers from the first item
        headers = list(data[0].keys()) if data else []
        
        html = f"""
        <!DOCTYPE html>
        <html>
        <head>
            <title>{title}</title>
            <style>
                body {{ font-family: Arial, sans-serif; margin: 20px; }}
                table {{ border-collapse: collapse; width: 100%; }}
                th, td {{ border: 1px solid #ddd; padding: 8px; text-align: left; }}
                th {{ background-color: #f2f2f2; }}
                .high-capacity {{ background-color: #ffebee; }}
                .critical-capacity {{ background-color: #ffcdd2; }}
                h1 {{ color: #333; }}
                .report-meta {{ color: #666; font-size: 0.9em; margin-bottom: 20px; }}
            </style>
        </head>
        <body>
            <h1>{title}</h1>
            <div class="report-meta">Generated on: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}</div>
            <table>
                <thead>
                    <tr>
                        {"".join(f"<th>{header.replace('_', ' ').title()}</th>" for header in headers)}
                    </tr>
                </thead>
                <tbody>
        """
        
        for row in data:
            # Add CSS class based on utilization percentage if available
            css_class = ""
            if 'utilizationPercentage' in row:
                if row['utilizationPercentage'] >= 95:
                    css_class = ' class="critical-capacity"'
                elif row['utilizationPercentage'] >= 90:
                    css_class = ' class="high-capacity"'
            
            html += f'<tr{css_class}>'
            for header in headers:
                value = row.get(header, "")
                if isinstance(value, float):
                    value = f"{value:.2f}"
                html += f"<td>{value}</td>"
            html += "</tr>"
        
        html += """
                </tbody>
            </table>
        </body>
        </html>
        """
        
        return html

File: static/js/reports.js
// Reports Dashboard JavaScript
class ReportsManager {
  constructor() {
    this.currentData = {};
    this.charts = {};
    this.init();
  }

  init() {
    // Load initial dashboard data
    this.loadDashboardData();

    // Set up event listeners
    this.setupEventListeners();
  }

  setupEventListeners() {
    // Tab switching
    document.querySelectorAll(".tab-btn").forEach((btn) => {
      btn.addEventListener("click", (e) => {
        const tabName = e.target
          .getAttribute("onclick")
          .match(/showTab\('(.+)'\)/)[1];
        this.showTab(tabName);
      });
    });
  }

  showTab(tabName) {
    // Remove active class from all tabs and content
    document
      .querySelectorAll(".tab-btn")
      .forEach((btn) => btn.classList.remove("active"));
    document
      .querySelectorAll(".tab-content")
      .forEach((content) => content.classList.remove("active"));

    // Add active class to selected tab and content
    document
      .querySelector(`[onclick="showTab('${tabName}')"]`)
      .classList.add("active");
    document.getElementById(tabName).classList.add("active");

    // Load data for the selected tab if needed
    this.loadTabData(tabName);
  }

  showLoading() {
    document.getElementById("loadingSpinner").style.display = "flex";
  }

  hideLoading() {
    document.getElementById("loadingSpinner").style.display = "none";
  }

  async loadDashboardData() {
    this.showLoading();
    try {
      const response = await fetch("/api/reports/dashboard");
      const result = await response.json();

      if (result.status === "Success") {
        this.currentData = result.data;
        this.updateOverviewMetrics();
        this.createCharts();
      } else {
        console.error("Failed to load dashboard data:", result.message);
      }
    } catch (error) {
      console.error("Error loading dashboard data:", error);
    } finally {
      this.hideLoading();
    }
  }

  updateOverviewMetrics() {
    const data = this.currentData;

    // Calculate metrics
    const totalCourses = data.enrollmentStatistics?.length || 0;
    const totalEnrollments =
      data.enrollmentStatistics?.reduce(
        (sum, course) => sum + course.filledSeats,
        0
      ) || 0;
    const avgUtilization =
      data.departmentAnalytics?.reduce(
        (sum, dept) => sum + dept.avgUtilization,
        0
      ) / (data.departmentAnalytics?.length || 1) || 0;
    const highCapacityCount = data.highCapacityCourses?.length || 0;

    // Update DOM elements
    document.getElementById("totalCourses").textContent = totalCourses;
    document.getElementById("totalEnrollments").textContent = totalEnrollments;
    document.getElementById(
      "avgUtilization"
    ).textContent = `${avgUtilization.toFixed(1)}%`;
    document.getElementById("highCapacityCount").textContent =
      highCapacityCount;
  }

  createCharts() {
    this.createDepartmentChart();
    this.createPopularityChart();
  }

  createDepartmentChart() {
    const ctx = document.getElementById("departmentChart").getContext("2d");
    const data = this.currentData.departmentAnalytics || [];

    if (this.charts.department) {
      this.charts.department.destroy();
    }

    this.charts.department = new Chart(ctx, {
      type: "bar",
      data: {
        labels: data.map((dept) => dept.departmentName),
        datasets: [
          {
            label: "Utilization %",
            data: data.map((dept) => dept.avgUtilization),
            backgroundColor: "rgba(37, 99, 235, 0.7)",
            borderColor: "rgba(37, 99, 235, 1)",
            borderWidth: 1,
          },
        ],
      },
      options: {
        responsive: true,
        scales: {
          y: {
            beginAtZero: true,
            max: 100,
          },
        },
      },
    });
  }

  createPopularityChart() {
    const ctx = document.getElementById("popularityChart").getContext("2d");
    const data = this.currentData.popularCourses || [];

    if (this.charts.popularity) {
      this.charts.popularity.destroy();
    }

    this.charts.popularity = new Chart(ctx, {
      type: "doughnut",
      data: {
        labels: data.map((course) => course.courseName),
        datasets: [
          {
            data: data.map((course) => course.enrolledCount),
            backgroundColor: [
              "#2563eb",
              "#10b981",
              "#f59e0b",
              "#ef4444",
              "#8b5cf6",
              "#06b6d4",
              "#84cc16",
              "#f97316",
              "#ec4899",
              "#6366f1",
            ],
          },
        ],
      },
      options: {
        responsive: true,
        plugins: {
          legend: {
            position: "bottom",
          },
        },
      },
    });
  }

  loadTabData(tabName) {
    switch (tabName) {
      case "enrollment":
        this.loadEnrollmentStats();
        break;
      case "faculty":
        this.loadFacultyWorkload();
        break;
      case "popularity":
        this.loadCoursePopularity();
        break;
      case "capacity":
        this.loadHighCapacityCourses();
        break;
      case "departments":
        this.loadDepartmentAnalytics();
        break;
    }
  }

  async loadEnrollmentStats() {
    const departmentId = document.getElementById(
      "enrollmentDepartmentFilter"
    ).value;

    this.showLoading();
    try {
      const url = departmentId
        ? `/api/reports/enrollment-statistics?department_id=${departmentId}`
        : "/api/reports/enrollment-statistics";

      const response = await fetch(url);
      const result = await response.json();

      if (result.status === "Success") {
        this.displayEnrollmentStatsTable(result.data);
      } else {
        console.error("Failed to load enrollment statistics:", result.message);
      }
    } catch (error) {
      console.error("Error loading enrollment statistics:", error);
    } finally {
      this.hideLoading();
    }
  }

  displayEnrollmentStatsTable(data) {
    const container = document.getElementById("enrollmentStatsTable");

    if (!data || data.length === 0) {
      container.innerHTML = "<p>No enrollment data available.</p>";
      return;
    }

    const table = this.createTable(
      [
        "Course Name",
        "Department",
        "Instructor",
        "Capacity",
        "Filled",
        "Available",
        "Utilization %",
        "Status",
      ],
      data.map((course) => [
        course.courseName,
        course.department,
        course.instructor,
        course.capacity,
        course.filledSeats,
        course.availableSeats,
        this.createUtilizationBar(course.utilizationPercentage),
        this.createStatusBadge(course.status),
      ])
    );

    container.innerHTML = "";
    container.appendChild(table);
  }

  async loadFacultyWorkload() {
    this.showLoading();
    try {
      const response = await fetch("/api/reports/faculty-workload");
      const result = await response.json();

      if (result.status === "Success") {
        this.displayFacultyWorkloadTable(result.data);
      } else {
        console.error("Failed to load faculty workload:", result.message);
      }
    } catch (error) {
      console.error("Error loading faculty workload:", error);
    } finally {
      this.hideLoading();
    }
  }

  displayFacultyWorkloadTable(data) {
    const container = document.getElementById("facultyWorkloadTable");

    if (!data || data.length === 0) {
      container.innerHTML = "<p>No faculty workload data available.</p>";
      return;
    }

    const table = this.createTable(
      [
        "Faculty Name",
        "Department",
        "Number of Courses",
        "Total Students",
        "Avg Students per Course",
      ],
      data.map((faculty) => [
        faculty.facultyName,
        faculty.department,
        faculty.numberOfCourses,
        faculty.totalEnrolledStudents,
        faculty.avgStudentsPerCourse,
      ])
    );

    container.innerHTML = "";
    container.appendChild(table);
  }

  async loadCoursePopularity() {
    const limit = document.getElementById("popularityLimit").value;

    this.showLoading();
    try {
      const response = await fetch(
        `/api/reports/course-popularity?limit=${limit}`
      );
      const result = await response.json();

      if (result.status === "Success") {
        this.displayCoursePopularityTable(result.data);
      } else {
        console.error("Failed to load course popularity:", result.message);
      }
    } catch (error) {
      console.error("Error loading course popularity:", error);
    } finally {
      this.hideLoading();
    }
  }

  displayCoursePopularityTable(data) {
    const container = document.getElementById("coursePopularityTable");

    if (!data || data.length === 0) {
      container.innerHTML = "<p>No course popularity data available.</p>";
      return;
    }

    const table = this.createTable(
      [
        "Rank",
        "Course Name",
        "Department",
        "Instructor",
        "Enrolled",
        "Capacity",
        "Popularity %",
      ],
      data.map((course) => [
        course.rank,
        course.courseName,
        course.department,
        course.instructor,
        course.enrolledCount,
        course.capacity,
        `${course.popularityPercentage}%`,
      ])
    );

    container.innerHTML = "";
    container.appendChild(table);
  }

  async loadHighCapacityCourses() {
    const department = document.getElementById(
      "capacityDepartmentFilter"
    ).value;
    const threshold = document.getElementById("capacityThreshold").value;

    this.showLoading();
    try {
      const params = new URLSearchParams();
      if (department) params.append("department", department);
      if (threshold) params.append("threshold", threshold);

      const response = await fetch(
        `/api/reports/high-capacity-courses?${params}`
      );
      const result = await response.json();

      if (result.status === "Success") {
        this.displayHighCapacityTable(result.data);
      } else {
        console.error("Failed to load high capacity courses:", result.message);
      }
    } catch (error) {
      console.error("Error loading high capacity courses:", error);
    } finally {
      this.hideLoading();
    }
  }

  displayHighCapacityTable(data) {
    const container = document.getElementById("highCapacityTable");

    if (!data || data.length === 0) {
      container.innerHTML = "<p>No high capacity courses found.</p>";
      return;
    }

    const table = this.createTable(
      [
        "Course Name",
        "Department",
        "Instructor",
        "Capacity",
        "Enrolled",
        "Utilization %",
        "Status",
      ],
      data.map((course) => [
        course.courseName,
        course.department,
        course.instructor,
        course.capacity,
        course.enrolledCount,
        this.createUtilizationBar(course.utilizationPercentage),
        this.createStatusBadge(course.status),
      ])
    );

    container.innerHTML = "";
    container.appendChild(table);
  }

  async loadDepartmentAnalytics() {
    this.showLoading();
    try {
      const response = await fetch("/api/reports/department-analytics");
      const result = await response.json();

      if (result.status === "Success") {
        this.displayDepartmentAnalyticsTable(result.data);
      } else {
        console.error("Failed to load department analytics:", result.message);
      }
    } catch (error) {
      console.error("Error loading department analytics:", error);
    } finally {
      this.hideLoading();
    }
  }

  displayDepartmentAnalyticsTable(data) {
    const container = document.getElementById("departmentAnalyticsTable");

    if (!data || data.length === 0) {
      container.innerHTML = "<p>No department analytics data available.</p>";
      return;
    }

    const table = this.createTable(
      [
        "Department",
        "Total Courses",
        "Total Capacity",
        "Total Enrolled",
        "Available Seats",
        "Avg Utilization %",
        "Faculty Count",
      ],
      data.map((dept) => [
        dept.departmentName,
        dept.totalCourses,
        dept.totalCapacity,
        dept.totalEnrolled,
        dept.totalAvailable,
        `${dept.avgUtilization}%`,
        dept.facultyCount,
      ])
    );

    container.innerHTML = "";
    container.appendChild(table);
  }

  async loadBusinessSchoolReport() {
    this.showLoading();
    try {
      const response = await fetch("/api/reports/business-school-capacity");
      const result = await response.json();

      if (result.status === "Success") {
        this.displayBusinessSchoolReport(result);
      } else {
        console.error("Failed to load business school report:", result.message);
      }
    } catch (error) {
      console.error("Error loading business school report:", error);
    } finally {
      this.hideLoading();
    }
  }

  displayBusinessSchoolReport(result) {
    const container = document.getElementById("highCapacityTable");
    const { data, summary } = result;

    if (!data || data.length === 0) {
      container.innerHTML =
        "<p>No Business school courses found above the capacity threshold.</p>";
      return;
    }

    // Add summary information
    const summaryHtml = `
            <div class="report-summary" style="margin-bottom: 1rem; padding: 1rem; background: #f8fafc; border-radius: 0.5rem;">
                <h4>Business School High Capacity Report Summary</h4>
                <div style="display: grid; grid-template-columns: repeat(auto-fit, minmax(150px, 1fr)); gap: 1rem; margin-top: 0.5rem;">
                    <div><strong>Department:</strong> ${summary.department}</div>
                    <div><strong>Threshold:</strong> ${summary.threshold}%</div>
                    <div><strong>Total Courses:</strong> ${summary.totalCourses}</div>
                    <div><strong>Avg Utilization:</strong> ${summary.averageUtilization}%</div>
                </div>
            </div>
        `;

    const table = this.createTable(
      [
        "Course Name",
        "Instructor",
        "Total Capacity",
        "Enrolled Students",
        "Utilization %",
        "Available Seats",
        "Status",
      ],
      data.map((course) => [
        course.courseName,
        course.instructor,
        course.totalCapacity,
        course.enrolledStudents,
        this.createUtilizationBar(course.utilizationPercentage),
        course.availableSeats,
        this.createStatusBadge(course.status),
      ])
    );

    container.innerHTML = summaryHtml;
    container.appendChild(table);
  }

  createTable(headers, rows) {
    const table = document.createElement("table");
    table.className = "report-table";

    // Create header
    const thead = document.createElement("thead");
    const headerRow = document.createElement("tr");
    headers.forEach((header) => {
      const th = document.createElement("th");
      th.textContent = header;
      headerRow.appendChild(th);
    });
    thead.appendChild(headerRow);
    table.appendChild(thead);

    // Create body
    const tbody = document.createElement("tbody");
    rows.forEach((row) => {
      const tr = document.createElement("tr");
      row.forEach((cell) => {
        const td = document.createElement("td");
        if (typeof cell === "string") {
          td.textContent = cell;
        } else {
          td.appendChild(cell);
        }
        tr.appendChild(td);
      });
      tbody.appendChild(tr);
    });
    table.appendChild(tbody);

    return table;
  }

  createUtilizationBar(percentage) {
    const container = document.createElement("div");
    container.style.display = "flex";
    container.style.alignItems = "center";
    container.style.gap = "0.5rem";

    const bar = document.createElement("div");
    bar.className = "utilization-bar";

    const fill = document.createElement("div");
    fill.className = "utilization-fill";
    fill.style.width = `${percentage}%`;

    if (percentage < 70) fill.classList.add("low");
    else if (percentage < 90) fill.classList.add("medium");
    else fill.classList.add("high");

    bar.appendChild(fill);

    const text = document.createElement("span");
    text.textContent = `${percentage}%`;
    text.style.fontSize = "0.75rem";
    text.style.fontWeight = "600";

    container.appendChild(bar);
    container.appendChild(text);

    return container;
  }

  createStatusBadge(status) {
    const badge = document.createElement("span");
    badge.className = "status-badge";
    badge.textContent = status;

    switch (status.toLowerCase()) {
      case "open":
        badge.classList.add("status-open");
        break;
      case "full":
        badge.classList.add("status-full");
        break;
      case "critical":
        badge.classList.add("status-critical");
        break;
      case "high":
      case "high capacity":
        badge.classList.add("status-high");
        break;
    }

    return badge;
  }

  async exportReport(reportType, format) {
    this.showLoading();
    try {
      const url = `/api/reports/export/${format}?type=${reportType}`;
      const response = await fetch(url);

      if (response.ok) {
        const blob = await response.blob();
        const downloadUrl = window.URL.createObjectURL(blob);
        const a = document.createElement("a");
        a.href = downloadUrl;
        a.download = `${reportType}_report.${format}`;
        document.body.appendChild(a);
        a.click();
        document.body.removeChild(a);
        window.URL.revokeObjectURL(downloadUrl);
      } else {
        console.error("Failed to export report");
      }
    } catch (error) {
      console.error("Error exporting report:", error);
    } finally {
      this.hideLoading();
    }
  }

  // Quick action methods
  async generateBusinessReport() {
    this.showTab("capacity");
    await this.loadBusinessSchoolReport();
  }

  async generateHighCapacityReport() {
    this.showTab("capacity");
    document.getElementById("capacityThreshold").value = "90";
    await this.loadHighCapacityCourses();
  }

  async generateFacultyReport() {
    this.showTab("faculty");
    await this.loadFacultyWorkload();
  }
}

// Global functions for template onclick handlers
function showTab(tabName) {
  window.reportsManager.showTab(tabName);
}

function loadEnrollmentStats() {
  window.reportsManager.loadEnrollmentStats();
}

function loadFacultyWorkload() {
  window.reportsManager.loadFacultyWorkload();
}

function loadCoursePopularity() {
  window.reportsManager.loadCoursePopularity();
}

function loadHighCapacityCourses() {
  window.reportsManager.loadHighCapacityCourses();
}

function loadBusinessSchoolReport() {
  window.reportsManager.loadBusinessSchoolReport();
}

function loadDepartmentAnalytics() {
  window.reportsManager.loadDepartmentAnalytics();
}

function exportReport(reportType, format) {
  window.reportsManager.exportReport(reportType, format);
}

function generateBusinessReport() {
  window.reportsManager.generateBusinessReport();
}

function generateHighCapacityReport() {
  window.reportsManager.generateHighCapacityReport();
}

function generateFacultyReport() {
  window.reportsManager.generateFacultyReport();
}

function closeExportModal() {
  document.getElementById("exportModal").style.display = "none";
}

// Initialize the reports manager when the page loads
document.addEventListener("DOMContentLoaded", function () {
  window.reportsManager = new ReportsManager();
});
